gpu4pyscf/nac/tdrks.py

```diff
@@ -0,0 +1,346 @@
+    """
+    if singlet is False:
+        raise NotImplementedError('Only supports for singlet states')
+    mol = td_nac.mol
+    mf = td_nac.base._scf
+    mf_grad = mf.nuc_grad_method()
+    mo_coeff = cp.asarray(mf.mo_coeff)
+    mo_energy = cp.asarray(mf.mo_energy)
+    mo_occ = cp.asarray(mf.mo_occ)
+    nao, nmo = mo_coeff.shape
+    nocc = int((mo_occ > 0).sum())
+    nvir = nmo - nocc
+    orbv = mo_coeff[:, nocc:]
+    orbo = mo_coeff[:, :nocc]
+
+    xI, yI = x_yI
+    xI = cp.asarray(xI).reshape(nocc, nvir).T
+    if not isinstance(yI, np.ndarray) and not isinstance(yI, cp.ndarray):
+        yI = cp.zeros_like(xI)
+    yI = cp.asarray(yI).reshape(nocc, nvir).T
+    LI = xI-yI    # eq.(83) in Ref. [1]
+
+    ni = mf._numint
+    ni.libxc.test_deriv_order(mf.xc, 3, raise_error=True)
+    omega, alpha, hyb = ni.rsh_and_hybrid_coeff(mf.xc, mol.spin)
+    with_k = ni.libxc.is_hybrid_xc(mf.xc)
+
+    vresp = mf.gen_response(singlet=None, hermi=1)
+
+    def fvind(x):
+        dm = reduce(cp.dot, (orbv, x.reshape(nvir, nocc) * 2, orbo.T)) # double occupency
+        v1ao = vresp(dm + dm.T)
+        return reduce(cp.dot, (orbv.T, v1ao, orbo)).ravel()
+
+    z1 = cphf.solve(
+        fvind,
+        mo_energy,
+        mo_occ,
+        -LI*1.0*EI, # only one spin, negative in cphf
+        max_cycle=td_nac.cphf_max_cycle,
+        tol=td_nac.cphf_conv_tol)[0] # eq.(83) in Ref. [1]
+
+    z1 = z1.reshape(nvir, nocc)
+    z1ao = reduce(cp.dot, (orbv, z1, orbo.T)) * 2 # double occupency
+    # eq.(50) in Ref. [1]
+    z1aoS = (z1ao + z1ao.T)*0.5 # 0.5 is in the definition of z1aoS
+    # eq.(73) in Ref. [1]
+    GZS = vresp(z1aoS) # generate the double occupency 
+    GZS_mo = reduce(cp.dot, (mo_coeff.T, GZS, mo_coeff))
+    W = cp.zeros((nmo, nmo))  # eq.(75) in Ref. [1]
+    W[:nocc, :nocc] = GZS_mo[:nocc, :nocc]
+    zeta0 = mo_energy[nocc:, cp.newaxis]
+    zeta0 = z1 * zeta0
+    W[:nocc, nocc:] = GZS_mo[:nocc, nocc:] + 0.5*yI.T*EI + 0.5*zeta0.T #* eq.(43), (56), (28) in Ref. [1]
+    zeta1 = mo_energy[cp.newaxis, :nocc]
+    zeta1 = z1 * zeta1
+    W[nocc:, :nocc] = 0.5*xI*EI + 0.5*zeta1
+    W = reduce(cp.dot, (mo_coeff, W , mo_coeff.T)) * 2.0
+
+    mf_grad = mf.nuc_grad_method()
+    s1 = mf_grad.get_ovlp(mol)
+    dmz1doo = z1aoS
+    oo0 = reduce(cp.dot, (orbo, orbo.T)) * 2.0
+
+    if atmlst is None:
+        atmlst = range(mol.natm)
+    
+    h1 = cp.asarray(mf_grad.get_hcore(mol))  # without 1/r like terms
+    s1 = cp.asarray(mf_grad.get_ovlp(mol))
+    dh_td = contract("xij,ij->xi", h1, dmz1doo)
+    ds = contract("xij,ij->xi", s1, (W + W.T))
+
+    dh1e_td = int3c2e.get_dh1e(mol, dmz1doo)  # 1/r like terms
+    if mol.has_ecp():
+        dh1e_td += rhf_grad.get_dh1e_ecp(mol, dmz1doo)  # 1/r like terms
+
+    j_factor = 1.0
+    k_factor = 0.0
+    if with_k:
+        k_factor = hyb
+    extra_force = cp.zeros((len(atmlst), 3))
+    dvhf_all = 0
+    dvhf = td_nac.get_veff(mol, dmz1doo + oo0, j_factor, k_factor) 
```

Comment:
The get_veff here is misleading as it actually computes only the JK part of the Veff. Modify the get_veff function (see comments in that function) to compute the actual Veff derivatives  or  remove the `get_veff` method from the NAC class and change this line to `nac.tdrhf.NAC.get_veff(td_nac, ...)`